### ISSUE TITLE
Argoexec for windows containers

### ISSUE THREAD

**Issue by gatepoet** [argoproj/argo-workflows#1507]
**Is this a BUG REPORT or FEATURE REQUEST?**: Feature request

Argoexec doesn't run on windows containers. This causes windows workflow steps to fail if e.g. artifacts are used. Due to Argoexec being written in Go, there should be a way to build a windows image and sew it together with the linux image into a multi platform image. I have already done this succsessfully.

The current problem with this is that there are a few places in the codebase that have hardcoded references to parts only available on linux systems. This fails a build when targeting windows. For our personal use, I have made such an image. It was surprisingly easy once I got to know the codebase a little. Basically, there are to calls in the pns.go executor, to [enter](https://github.com/argoproj/argo/blob/master/workflow/executor/pns/pns.go#L99) and [exit](https://github.com/argoproj/argo/blob/master/workflow/executor/pns/pns.go#L111) chroot. I have deleted these lines in my build. There is also a [reference to SIGUSR2](https://github.com/argoproj/argo/blob/master/workflow/executor/executor.go#L957) in executor.go which I have changed to SIGINT to pass the build. I'm not sure how this would affect the execution of workflows, but the code says something about updating pod annotations faster.

I'm currently building using the following Dockerfile
```Dockerfile
FROM mcr.microsoft.com/windows/servercore:1809 as core
FROM mcr.microsoft.com/windows/nanoserver:1809 as argoexec-base
COPY --from=core /windows/system32/netapi32.dll /windows/system32/netapi32.dll

# NOTE: keep the version synced with https://storage.googleapis.com/kubernetes-release/release/stable.txt
ENV KUBECTL_VERSION=1.14.3
ADD https://storage.googleapis.com/kubernetes-release/release/v${KUBECTL_VERSION}/bin/windows/amd64/kubectl.exe /app/

USER Administrator
RUN SETX /m path C:\app;%path%

FROM argoexec-base
COPY dist/argoexec /app/argoexec.exe
```

and the following YAML to make the multi-platform image using 'manifest-tool'
```yaml
image: my-personal-repo/argoexec:latest
manifests:
  -
    image: my-personal-repo/argoexec:latest-linux-amd64
    platform:
      architecture: amd64
      os: linux
  -
    image: my-personal-repo/argoexec:latest-windows-amd64
    platform:
      architecture: amd64
      os: windows
```
Note that I use my personal image repo where I have pulled the current argoexec image and retagged it, appending `-linux-amd64`

Building this way will allow you to use windows container steps with artifact passing. The multi-platform image also allows for workflow with both linux and windows steps.

So, even though this is relatively easy to do locally, I have no clear suggestion on how you would incorporate this into your build process. Ideally we'd refactor all the platform specific features into a substitutable/optional dependency.

But the amount of business value provided by being able to run windows container steps is so huge, that I hope you're able to somewhat prioritize this. I'd make you a pull request, but I don't feel like I know the codebase or the Go programming language well enough.

**Comment by jessesuen** [comment]
First off, I love the idea of windows support for workflows.

> So, even though this is relatively easy to do locally, I have no clear suggestion on how you would incorporate this into your build process. Ideally we'd refactor all the platform specific features into a substitutable/optional dependency.

To really accomplish this, I think we would need to switch to something like Circle CI for the build process.

**Comment by crabulik** [comment]
@jessesuen Is there any information about Argoexec support for Windows? Will it be somewhere in the nearest future?

**Comment by lippertmarkus** [comment]
I implemented this in #2747 if you want to have a look. This issue motivated me to do so :)
